dnprogs/dnsubmit/dnsubmit.1
.TH DNDEL 1 "October 2 1998" "DECnet utilities"

.SH NAME
dnsubmit \- Submit a batch job on a VMS system
.br
dnprint \- Print a file on a VMS system
.SH SYNOPSIS
.B dnsubmit
[options] file-name
.br
.B dnprint
[options] file-name
.br
Options:
.br
[\-vh]
.SH DESCRIPTION
.PP
.B dnprint 
and
.B dnsubmit
send files to a VMS machine for printing or batch execution.
These command behave similarly to the VMS PRINT/REMOTE and SUBMIT/REMOTE
commands in that you get very little control of the queueing mechanism.
.br
dnprint always prints to the queue SYS$PRINT and dnsubmit always submits
to SYS$BATCH. Of course you can always redirect these queues using
logical names.
.br
See the man page for 
.B dncopy 
for a discussion of VMS file specifications.

.SH OPTIONS
.TP
.I "\-T connect timeout"
Specifies the maximum amount of time the command will wait to establish a connection
with the remote node. a 0 here will cause it to wait forever. The default is 60 seconds
.TP
.I \-h \-?
Displays help for using the command.
.TP
.I \-V
Show the version of the tools package that the program comes from.

.SH EXAMPLES

  dnsubmit 'myvax::myjob.com'

.br
  dnprint 'tramp"christine pjc123"::file.lis'

.SH SEE ALSO
.BR dntype "(1), " dndir "(1), " dncopy "(1), " dntask "(1), " dndel "(1)"


dnprogs/NEWS
dnprogs (2.61) unstable; urgency=low

  * Include libvaxdata/src/test.c in the distribution tarball
  * Fix password length in DAP connections. It was incorrectly set at 12 
    rather than 40
 
 -- Chrissie Caulfield <chrissie@debian.org>  Sun, 31 Mar 2013 11:18:32 +0000

dnprogs (2.60+x32) unreleased; urgency=low

  * Switch from using deprecated sysctl(2) to writing to /proc/sys.  This
    fixes a build failure on x32.

 -- Daniel Schepler <schepler@debian.org>  Tue, 12 Feb 2013 12:21:31 -0800

dnprogs (2.60) unstable; urgency=low

  * Fixed lots of Debian things that caused lintian problems and 
    upgraded policy version
    Closes: #673408
  * Fix some warnings that might be bugs
  * Incorporate NMU into main tree.

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 26 May 2012 09:54:41 +0100

dnprogs (2.59+nmu2) unstable; urgency=medium

  * NMU
  * dnprogs binaries are Linux-only, so state that. Move dnet-common back
    to binary-all. Closes: #665886

 -- Steve McIntyre <93sam@debian.org>  Mon, 21 May 2012 17:43:14 +0100

dnprogs (2.59+nmu1) unstable; urgency=low

  * NMU
  * Reduce libdnet's recommends on dnet-common to suggests again.
    Closes: #655740. Adapted patch from Jonathan Nieder.

 -- Steve McIntyre <93sam@debian.org>  Fri, 18 May 2012 13:29:53 +0100

dnprogs (2.59) unstable; urgency=low

  * Changed package dnet-common from arch all to any and
    updated dependecies. (Closes: #665886)

 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 27 Apr 2012 09:30:19 +0100

dnprogs (2.58) unstable; urgency=medium

  * Fixed lintian warnings:
    - Fixed changelog entry of last release to have correct line length.
    - Corrected debconf dependency in dnet-common.
    - Corrected versioned dependency in dnprogs on dnet-common.
    - Updated existing overrides to match new syntax.
  * Updated linkage of libdnet_daemon and libdap. (Closes: #646353)
  * Fixed compile warning in dndir when using -Werror=format-security
    (Closes: #646613)
  * Fixed typo in debian/dnet-common.config
  * Do not try to install files into packages on FreeBSD which are
    not build on FreeBSD.
  * Link dapfs with libpthread for new libfuse compatibility
  * Updated nl.po (Closes: #652349)
  * Added pl.po (Closes: #661894)
  * Added it.po (Closes: #663135)
  * Change module-init-tools dependency to kmod (Closes: #662684)

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 10 Mar 2012 10:32:08 +0000

dnprogs (2.57) unstable; urgency=medium

  [ Chrissie Caulfield ]
  * Completed removal of modutils configuration file code
  * Fix presence of CVS directories in tarball
  * updated 'clean' target

  [ Philipp Schafft ]
  * Use /dev/pts PTYs on 64 bit systems (fix makefile bug)
  * Updated translations Debconf templates:
    - German translation (Closes: #615268)
    - French translation (Closes: #615039)
    - Danish translation (Closes: #633984) (added)
    - Slovak translation (Closes: #639629) (added)
  * Look in yet another place for libcrypt (better fix than from NMUs)
  * Added a comment about the detecting of libcrypt. (See: #629664)
  * "ethernet" was capitalized in extended description (Closes: #636798)
  * Fixed some lintian warnings (mainly new rules targets).
  * dnet-common: Do not configure DECnet unless the user requested it.
    (Removed the bad workaround done by NMU) (See: #637179)
  * dnet-common: Added preinst script to clean up the broken config files
    from the last NMU. (Closes: #640861)
  * Upgraded libdnet's suggest to dnet-common to recommend again
    (Closes: #608807)

 -- Chrissie Caulfield <chrissie@debian.org>  Thu,  8 Sep 2011 16:05:47 +0200

dnprogs (2.56.1+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * libdnet: Only suggest dnet-common package (Closes: #608807)
  * dnet-common: Do not change MAC addresses unless user chooses to
    configure DECnet (Closes: #637179)
  * dnet-common: Restore ability to configure DECnet name and address
    through debconf
  * dnet-common: Remove obsolete modutils configuration file

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 25 Aug 2011 14:26:12 +0100

dnprogs (2.56.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add patch from Ubuntu to fix FTBFS with libcrypt moved to the
    multiarch path.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 09 Aug 2011 11:58:06 +0200


dnprogs (2.56) unstable; urgency=low

  * Search for libgcrypt in 64bit library paths, too. (thanks to Kamil)
  * Package dnet-common now asks if network interfaces should be configrued
    for DECnet (Closes: #608807)
    Updated translations Debconf templates:
    - Russian translation (Closes: #610642)
    - Portuguese translation (Closes: #611034)
    - Swedish translation (Closes: #611296)
    - Czech translation (Closes: #611430)
    - Japanese translation (Closes: #611448)
    - Spanish translation (Closes: #611547)
  * Build against version 25 of FUSE
    Closes: #598831
  * Fixed some lintian warnings (typos and such).
  * Raised priority of package libdnet from extra to optional
    to comply policy manual section 2.5 "Priorities".

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 22 Feb 2011 10:38:23 +0000

dnprogs (2.55) unstable; urgency=low

  * Build against version 25 of FUSE
    Closes: #598831

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 5 Oct 2010 09:23:12 +0100

dnprogs (2.54) unstable; urgency=low

  * Don't build libvaxdata test program by default.
    Closes: #598752
 
 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 1 Oct 2010 20:44:09 +0100

dnprogs (2.53) unstable; urgency=low

  * Done some more porting to FreeBSD, corrected Build-Depends for hurd-i386
    Closes: #588185
  * Added new tool node(1) - Node name lookup tool
  * Added support to libdnet_daemon to read /etc/nodes.{allow,deny}
  * Fixed DoS bug in dnet_daemon()
    Closes: #592015 
  * Update libvaxdata to version 1.2, fixes double conversions

 -- Christine Caulfield <Christine.Caulfield@googlemail.com>  Thu, 19 Jul 2010 10:19:32 +0100


dnprogs/dnsubmit/dnsubmit.cc
/******************************************************************************
    (c) 1998-2008      Christine Caulfield          christine.caulfield@googlemail.com

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
 ******************************************************************************
 */
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <sys/types.h>
#include <string.h>
#include <errno.h>
#include <sys/socket.h>
#include <sys/ioctl.h>
#include <netdnet/dn.h>
#include <netdnet/dnetdb.h>
#include <unistd.h>
#include <regex.h>

#include "connection.h"
#include "protocol.h"
#include "logging.h"



/*-------------------------------------------------------------------------*/
static void usage(FILE *f, bool dnprint)
{
    if (dnprint)
    {
	fprintf(f,"\nUSAGE: dnprint [OPTIONS] 'node\"user password\"::filespec'\n\n");
    }
    else
    {
	fprintf(f, "\nUSAGE: dnsubmit [OPTIONS] 'node\"user password\"::filespec'\n\n");
    }
    fprintf(f,"NOTE: The VMS filename really should be in single quotes to\n");
    fprintf(f,"      protect it from the shell\n");

    fprintf(f,"\nOptions:\n");
    fprintf(f,"  -? -h        display this help message\n");
    fprintf(f,"  -T <secs>    connect timeout (default 60)\n");
    fprintf(f,"  -v           increase verbosity\n");
    fprintf(f,"  -V           show version number\n");

    fprintf(f,"\nExample:\n\n");
    if (dnprint)
    {
	fprintf(f," dnprint  'serv1\"user password\"::helloworld.lis'\n");
    }
    else
    {
	fprintf(f," dnsubmit  'serv1\"user password\"::myjob.com'\n");
    }
    fprintf(f,"\n");
}
/*-------------------------------------------------------------------------*/
static int submit(dap_connection &conn, char *dirname, bool print)
{
    dap_attrib_message att;
    dap_access_message acc;

    // When printing we set the SPOOL option on closing the file.
    if (print)
	acc.set_accfunc(dap_access_message::OPEN);
    else
	acc.set_accfunc(dap_access_message::SUBMIT);

    acc.set_filespec(dirname);

    conn.set_blocked(true);
    att.set_file("/", false);
    att.write(conn); // Send empty ATTRIB message
    acc.write(conn);
    return conn.set_blocked(false);
}
/*-------------------------------------------------------------------------*/
static int read_reply(dap_connection &conn)
{
    dap_message *m;

    m = dap_message::read_message(conn, true);
    if (!m) return -1;

    switch (m->get_type())
    {
    case dap_message::ATTRIB:
	break;

    case dap_message::ACK:
	break;

    case dap_message::STATUS:
        {
	    dap_status_message *sm = (dap_status_message *)m;
	    fprintf(stderr, "%s\n", sm->get_message());
	    return -1;
	}

    case dap_message::ACCOMP:
	break;

    case dap_message::NAME:
	break;

    default:
	printf("Unknown mesage received: 0x%x\n", m->get_type());
	return -1;
	break;
    }
    return 0;
}


int main(int argc, char *argv[])
{
    int	    opt,retval;
    int     verbose = 0;
    bool    dnprint = false;
    int     connect_timeout = 60;

    // Work out the command name
    if (strstr(argv[0], "dnprint"))
    {
	dnprint = true;
    }

    if (argc < 2)
    {
	usage(stderr, dnprint);
	exit(0);
    }

/* Get command-line options */
    opterr = 0;
    optind = 0;
    while ((opt=getopt(argc,argv,"?hvVT:")) != EOF)
    {
	switch(opt)
	{
	case 'h':
	    usage(stdout, dnprint);
	    exit(1);

	case '?':
	    usage(stderr, dnprint);
	    exit(1);

	case 'v':
	    verbose++;
	    break;

	case 'T':
	    connect_timeout = atoi(optarg);
	    break;

	case 'V':
	    printf("\ndnsubmit from dnprogs version %s\n\n", VERSION);
	    exit(1);
	    break;
	}
    }
    if (optind >= argc)
    {
	usage(stderr, dnprint);
	exit(2);
    }

    init_logging("dnsubmit", 'e', false);

    dap_connection conn(verbose);
    conn.set_connect_timeout(connect_timeout);

    char dirname[256] = {'\0'};
    if (!conn.connect(argv[optind], dap_connection::FAL_OBJECT, dirname))
    {
	fprintf(stderr, "%s\n", conn.get_error());
	return -1;
    }

    // Exchange config messages
    if (!conn.exchange_config())
    {
	fprintf(stderr, "Error in config: %s\n", conn.get_error());
	return -1;
    }

    if (!submit(conn, dirname, dnprint))
    {
	fprintf(stderr, "Error in opening: %s %s\n", dirname, conn.get_error());
	return -1;
    }

    retval = read_reply(conn);
    if (retval < 0) dnprint=false; // Don't try the rest

    // When printing we have to send an ACCOMP message with the SPL bit set.
    if (dnprint)
    {
	dap_accomp_message acc;

	acc.set_cmpfunc(dap_accomp_message::CLOSE);
	acc.set_fop_bit(dap_attrib_message::FB$SPL);
	acc.write(conn);

	retval = read_reply(conn);
    }

    conn.close();
    return 0;
}
